pkg/component/tls.go

```diff
@@ -50,31 +57,129 @@ func WithNextProtos(protos ...string) TLSConfigOption {
+)
+
+// GetTLSServerConfig gets the component's server TLS config and applies the given options.
+func (c *Component) GetTLSServerConfig(ctx context.Context, opts ...TLSConfigOption) (*tls.Config, error) {
+	var (
+		logger = log.FromContext(ctx)
+		conf   = c.GetBaseConfig(ctx).TLS
+		res    *tls.Config
+	)
+	for _, src := range []struct {
+		Enable            bool
+		CertificateGetter func() (func(*tls.ClientHelloInfo) (*tls.Certificate, error), error)
+	}{
+		{
+			Enable: conf.Certificate != "" && conf.Key != "",
+			CertificateGetter: func() (func(*tls.ClientHelloInfo) (*tls.Certificate, error), error) {
+				var cv atomic.Value
+				loadCertificate := func() error {
+					cert, err := tls.LoadX509KeyPair(conf.Certificate, conf.Key)
+					if err != nil {
+						return err
+					}
+					cv.Store(&cert)
+					logger.Debug("Loaded TLS certificate")
+					return nil
+				}
+				if err := loadCertificate(); err != nil {
+					return nil, err
 				}
-				return c.acme.GetCertificate(hello)
+				debounce := make(chan struct{}, 1)
+				fs.Watch(conf.Certificate, events.HandlerFunc(func(evt events.Event) {
+					if evt.Name() != "fs.write" {
+						return
+					}
+					// We have to debounce this; OpenSSL typically causes a lot of write events.
+					select {
+					case debounce <- struct{}{}:
+						time.AfterFunc(5*time.Second, func() {
+							if err := loadCertificate(); err != nil {
+								logger.WithError(err).Error("Could not reload TLS certificate")
+								return
+							}
+							<-debounce
+						})
+					default:
+					}
+				}))
+				return func(info *tls.ClientHelloInfo) (*tls.Certificate, error) {
+					return cv.Load().(*tls.Certificate), nil
+				}, nil
 			},
+		},
+		{
+			Enable: conf.ACME.Enable,
+			CertificateGetter: func() (func(*tls.ClientHelloInfo) (*tls.Certificate, error), error) {
+				opts = append(opts, WithNextProtos(acme.ALPNProto))
+				return func(hello *tls.ClientHelloInfo) (*tls.Certificate, error) {
+					if hello.ServerName == "" {
+						hello.ServerName = conf.ACME.DefaultHost
+					}
+					return c.acme.GetCertificate(hello)
+				}, nil
+			},
+		},
+		{
+			Enable: conf.KeyVault.Enable,
+			CertificateGetter: func() (func(*tls.ClientHelloInfo) (*tls.Certificate, error), error) {
+				if conf.KeyVault.ID == "" {
+					return nil, errTLSKeyVaultID
+				}
+				return func(*tls.ClientHelloInfo) (*tls.Certificate, error) {
+					return c.KeyVault.LoadCertificate(conf.KeyVault.ID)
+				}, nil
+			},
+		},
+	} {
+		if !src.Enable {
+			continue
+		}
+		if res != nil {
+			return nil, errAmbiguousTLSConfig
 		}
-		opts = append(opts, WithNextProtos(acme.ALPNProto))
-	} else {
-		var err error
-		conf, err = c.config.TLS.Config(ctx)
+		fn, err := src.CertificateGetter()
+		if err != nil {
+			return nil, err
+		}
+		res = &tls.Config{
+			GetCertificate: fn,
+		}
+	}
+	if res == nil {
+		return nil, errEmptyTLSConfig
+	}
+	res.MinVersion = tls.VersionTLS12
+	res.NextProtos = []string{"h2", "http/1.1"}
```

Comment:
I think this should at some point be moved to the listeners, because the MQTT listeners for example should't offer http2 or http1.1 in ALPN